subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1251 @@
+	uint8_t addr;			/* Endpoint address. */
+};
+
+struct backend_data {
+	const struct icmsg_with_buf_config *conf; /* Backend instance config. */
+	struct icmsg_data_t icmsg_data;	/* ICMsg data. */
+	struct k_mutex mutex;		/* Mutex to protect: ICMsg send call and
+					 * waiting_bound field.
+					 */
+	struct k_work ep_bound_work;	/* Work item for bounding processing. */
+	struct k_sem block_wait_sem;	/* Semaphore for waiting for free blocks. */
+	struct ept_data ept[NUM_EPT];	/* Array of registered endpoints. */
+	uint8_t ept_map[NUM_EPT];	/* Array that maps endpoint address to index. */
+	uint16_t waiting_bound[NUM_EPT];/* The bound messages waiting to be registered. */
+	atomic_t flags;			/* Flags on higher bits, number of registered
+					 * endpoints on lower.
+					 */
+	bool is_initiator;		/* This side has an initiator role. */
+};
+
+struct block_header {
+	volatile size_t size;	/* Size of the data field. It must be volatile, because
+				 * when this value is read and validated for security
+				 * reasons, compiler cannot generate code that reads
+				 * it again after validation.
+				 */
+	uint8_t data[];		/* Buffer data. */
+};
+
+struct icmsg_message {
+	uint8_t msg_type;	/* Message type. */
+	uint8_t ept_addr;	/* Endpoint address or zero for MSG_RELEASE_DATA. */
+	uint8_t block_index;	/* Block index to send or release. */
+};
+
+BUILD_ASSERT(NUM_EPT <= EPT_ADDR_INVALID, "Too many endpoints");
+
+/**
+ * Calculate pointer to block from its index and channel configuration (RX or TX).
+ * No validation is performed.
+ */
+static struct block_header *block_from_index(const struct channel_config *ch_conf,
+					     size_t block_index)
+{
+	return (struct block_header *)(ch_conf->blocks_ptr + block_index *
+				       ch_conf->block_size);
+}
+
+/**
+ * Calculate pointer to data buffer from block index and channel configuration (RX or TX).
+ * Also validate the index and optionally the buffer size allocated on the this block.
+ *
+ * @param[in]  ch_conf		The channel
+ * @param[in]  block_index	Block index
+ * @param[out] size		Size of the buffer allocated on the block if not NULL.
+ *				The size is also checked if it fits in the blocks area.
+ *				If it is NULL, no size validation is performed.
+ * @param[in]  invalidate_cache	If size is not NULL, invalidates cache for entire buffer
+ *				(all blocks). Otherwise, it is ignored.
+ * @return	Pointer to data buffer or NULL if validation failed.
+ */
+static uint8_t *buffer_from_index_validate(const struct channel_config *ch_conf,
+					   size_t block_index, size_t *size,
+					   bool invalidate_cache)
+{
+
+	size_t allocable_size;
+	size_t buffer_size;
+	uint8_t *end_ptr;
+	struct block_header *block;
+
+	if (block_index >= ch_conf->block_count) {
+		LOG_ERR("Block index invalid");
+		return NULL;
+	}
+
+	block = block_from_index(ch_conf, block_index);
+
+	if (size != NULL) {
+		if (invalidate_cache) {
+			sys_cache_data_invd_range(block, BLOCK_HEADER_SIZE);
```

Comment:
Assuming that:
1) the receiver does not write to received buffer (it is ensured on language level by using `const`)
2) the receiver does not flush the cache with received buffer (he don't need to, because he did not write anything)
then there is no need to align to the cache line.

That's a big if, so I implemented a cache aligned blocks that are configurable and enabled by default.